### PR TITLE
WW-5336 Deprecate OgnlTool

### PR DESCRIPTION
**Pull request by kusalk**
WW-5336
--
I couldn't see any value in retaining OgnlTool as its own class so I've deprecated it and merged the method into StrutsUtil.

### CODE REVIEW

**Comment by kusalk** [comment]
Await merge of #731 first

File: core/src/main/java/org/apache/struts2/util/StrutsUtil.java [hunk 1]
```diff
@@ -50,32 +61,30 @@ public class StrutsUtil {
 
     protected HttpServletRequest request;
     protected HttpServletResponse response;
-    protected Map<String, Class> classes = new Hashtable<>();
-    protected OgnlTool ognl;
+    protected Map<String, Class<?>> classes = new HashMap<>();
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/util/StrutsUtil.java:64]: This class is constructed per request thread so I don't believe we need concurrency handling here

File: core/src/main/java/org/apache/struts2/util/StrutsUtil.java [hunk 2]
```diff
@@ -50,32 +61,30 @@ public class StrutsUtil {
...
-        String name = aName.toString();
-        Class c = classes.get(name);
-
-        if (c == null) {
-            c = ClassLoaderUtil.loadClass(name, StrutsUtil.class);
-            classes.put(name, c);
+    public Object bean(Object name) throws Exception {
+        String className = name.toString();
+        Class<?> clazz = classes.get(className);
+        if (clazz == null) {
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/util/StrutsUtil.java:83]: Using #computeIfAbsent here would change the exception handling semantics so I've forgone it

File: core/src/main/java/org/apache/struts2/util/StrutsUtil.java [hunk 3]
```diff
@@ -124,6 +123,17 @@ public Object findValue(String expression, String className) throws ClassNotFoun
         return stack.findValue(expression, Class.forName(className));
     }
 
+    public Object findValue(String expr, Object context) {
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/util/StrutsUtil.java:126]: Copied this here from OgnlTool

File: core/src/main/java/org/apache/struts2/util/StrutsUtil.java [hunk 4]
```diff
@@ -124,6 +123,17 @@ public Object findValue(String expression, String className) throws ClassNotFoun
...
+        try {
+            return ognl.getValue(expr, ActionContext.getContext().getContextMap(), context);
+        } catch (OgnlException e) {
+            if (e.getReason() instanceof SecurityException) {
+                LOG.error(format("Could not evaluate this expression due to security constraints: [{0}]", expr), e);
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/util/StrutsUtil.java:131]: Fixed broken logging

File: core/src/main/java/org/apache/struts2/util/StrutsUtil.java [hunk 5]
```diff
@@ -156,71 +166,64 @@ public String translateVariables(String expression) {
...
-                selectedItems = Arrays.asList((Object[]) i);
-            } else if (i instanceof Collection) {
-                selectedItems = (Collection) i;
-            } else {
-                // treat it is a single item
-                selectedItems = new ArrayList();
-                selectedItems.add(i);
-            }
-        }
+    public List<ListEntry> makeSelectList(String selectedList, String list, String listKey, String listValue) {
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/util/StrutsUtil.java:169]: Fixed cognitive complexity and readability

File: core/src/main/resources/struts-beans.xml [hunk 6]
```diff
@@ -198,7 +198,7 @@
     <bean type="ognl.MethodAccessor" name="com.opensymphony.xwork2.util.CompoundRoot"
           class="com.opensymphony.xwork2.ognl.accessor.CompoundRootAccessor"/>
 
-    <bean class="org.apache.struts2.views.jsp.ui.OgnlTool"/>
+    <bean class="org.apache.struts2.views.jsp.ui.OgnlTool"/> <!-- Deprecated since 6.3.0 -->
```
Comment (sepe81) [inline core/src/main/resources/struts-beans.xml:201]: Maybe it's better to put the comment before the bean?

File: core/src/test/java/org/apache/struts2/util/StrutsUtilTest.java [hunk 7]
```diff
@@ -145,28 +147,51 @@ public void testGetContextMethod() {
...
+    public void testMakeSelectList() {
+        String[] selectedList = new String[]{"Car", "Airplane", "Bus"};
+        List<String> list = Arrays.asList("Lorry", "Car", "Helicopter");
+
+        stack.getContext().put("mySelectedList", selectedList);
+        stack.getContext().put("myList", list);
+
+        List<ListEntry> listMade = strutsUtil.makeSelectList("#mySelectedList", "#myList", null, null);
+
+        assertEquals(listMade.size(), 3);
```
Comment (sepe81) [inline core/src/test/java/org/apache/struts2/util/StrutsUtilTest.java:159]: for assertEquals, the expected value should be the first argument

**Review by sepe81 — CHANGES_REQUESTED** [review]

File: core/src/main/resources/struts-beans.xml (hunk 6 above)
Comment (kusalk) [reply core/src/main/resources/struts-beans.xml]: Amended - I tend to favour the same line as sometimes comments can become detached from their intended line in highly concurrent teams

File: core/src/test/java/org/apache/struts2/util/StrutsUtilTest.java (hunk 7 above)
Comment (kusalk) [reply core/src/test/java/org/apache/struts2/util/StrutsUtilTest.java]: Seems all the assertions in this class are mis-ordered, I've amended the ones I've touched :)

**Review by lukaszlenart — APPROVED** [review]
LGTM 👍

**Review by sepe81 — APPROVED** [review]

**Comment by lukaszlenart** [comment]
Hope this is the last change and I can prepare another RC or even test build and call for vote :)

**Pull request merged**